### PR TITLE
Add a decode try if returned sdb is bytes

### LINKED ISSUE #55360: Consul SDB treats all values as binary in Python 3
### Description of Issue
When using the Consul SDB module, all values retrieved from the KV store are rendered as binary strings, rather than unicode when using Python 3. They should be decoded to unicode, or use the appropriate function from six to allow for appropriate behavior between 2 and 3.

### Setup
In the master configuration:
```
    sdb:
      consul:
        driver: consul
        host: consul.service.consul
```

### Steps to Reproduce Issue
Set a value in the Consul SDB and then render it out into a salt template using Jinja to load the SDB value. e.g. `foo: {{ salt.sdb.get('sdb://consul/foo') }}`

### Versions Report
```
Salt Version:
           Salt: 2019.2.2
 
Dependency Versions:
           cffi: 1.13.2
       cherrypy: Not Installed
       dateutil: 2.5.3
      docker-py: Not Installed
          gitdb: 2.0.0
      gitpython: 2.1.1
          ioflo: Not Installed
         Jinja2: 2.8
        libgit2: 0.28.2
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: Not Installed
      pycparser: 2.19
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: 0.28.2
         Python: 3.5.3 (default, Sep 27 2018, 17:25:39)
   python-gnupg: Not Installed
         PyYAML: 3.12
          PyZMQ: 16.0.2
           RAET: Not Installed
          smmap: 2.0.1
        timelib: Not Installed
        Tornado: 4.4.3
            ZMQ: 4.2.1
 
System Versions:
           dist: debian 9.11 
         locale: UTF-8
        machine: x86_64
        release: 4.9.0-9-amd64
         system: Linux
        version: debian 9.11
```

### LINKED ISSUE #58533: [BUG] string values set with sdb.set come back as bytes from sdb.get
**Description**

String values set with `sdb.set` (using at least the redis or sqlite3 drivers) are retrieved as bytes.  Presumably, they should come back as strings.

**Setup**

Add the following as `/etc/salt/master.d/sdb.conf` and `/etc/salt/minion.d/sdb.conf`:

```yaml
sdb_redis:
  driver: redis
  host: redis
  port: 6379
  db: 1

sdb_sqlite:
  driver: sqlite3
  database: /tmp/sdb.sqlite
  table: sdb
  create_table: True
```

`/srv/salt/sdb-roundtrip.sls`:

```sls
{% for sdb in ['sdb_redis', 'sdb_sqlite'] %}
  {% set SDB_URI = 'sdb://{}/foo'.format(sdb) %}
  {% set SDB_VALUE = 'bar' %}

roundtrip_{{ sdb }}:
  test.configurable_test_state:
    - result: {{ salt.sdb.get(SDB_URI) == SDB_VALUE }}

roundtrip_{{ sdb }}-but-value-to-bytes:
  test.configurable_test_state:
    - result: {{ salt.sdb.get(SDB_URI) == SDB_VALUE|to_bytes }}

{% endfor %}{# sdb in ['sdb_redis', 'sdb_sqlite'] #}
```

**Steps to Reproduce the behavior**

The states with the original value converted to bytes match.

```text
# salt master1 --state-output terse state.apply zd5757.roundtrip
master1:
  Name: roundtrip_sdb_redis - Function: test.configurable_test_state - Result: Failed Started: - 18:26:49.549289 Duration: 1.759 ms
  Name: roundtrip_sdb_redis-but-value-to-bytes - Function: test.configurable_test_state - Result: Changed Started: - 18:26:49.558864 Duration: 1.129 ms
  Name: roundtrip_sdb_sqlite - Function: test.configurable_test_state - Result: Failed Started: - 18:26:49.566279 Duration: 1.029 ms
  Name: roundtrip_sdb_sqlite-but-value-to-bytes - Function: test.configurable_test_state - Result: Changed Started: - 18:26:49.573589 Duration: 1.012 ms
```

Or alternately:

```
[root@master1 /]# salt-call sdb.set sdb://sdb_redis/baz quux
local:
    True
[root@master1 /]# salt-call --out pprint sdb.get sdb://sdb_redis/baz 
{'local': b'quux'}
```

**Expected behavior**

`sdb.get` should return a string when set as a string.

**Versions Report**
<details><summary>salt --versions-report</summary>
(Provided by running salt --versions-report. Please also mention any differences in master/minion versions.) 

```
Salt Version:
           Salt: 3001.1
 
Dependency Versions:
           cffi: 1.14.3
       cherrypy: Not Installed
       dateutil: Not Installed
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
         Jinja2: 2.11.1
        libgit2: Not Installed
       M2Crypto: 0.35.2
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.6.2
   mysql-python: Not Installed
      pycparser: 2.20
       pycrypto: Not Installed
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 3.6.8 (default, Apr  2 2020, 13:34:55)
   python-gnupg: Not Installed
         PyYAML: 3.13
          PyZMQ: 17.0.0
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.5.3
            ZMQ: 4.1.4
 
System Versions:
           dist: centos 7 Core
         locale: UTF-8
        machine: x86_64
        release: 5.4.0-47-generic
         system: Linux
        version: CentOS Linux 7 Core
```
</details>

**Additional context**

ZD-5757.

### PR DESCRIPTION
**Pull request by johnvox**
### What does this PR do?
This PR fix an issue found on sdb get utils it not return always string and it produced side effect

### What issues does this PR fix or reference?
Fixes: #58533 
Fixes: #55360

### Previous Behavior
(on master config)
```
redis:
  driver: redis
  host: _redishost_
  password: _redispassword_
  port: _redisport_
  ssl: True

git_pillar_password: "sdb://redis/gitlab/password"
git_pillar_user: "sdb://redis/gitlab/user"
gitfs_remotes:
    - https://gitlab.com/<redacted>.git:
      - base: master
```
```
$ salt-master -l warning
[ERROR   ] Error occurred fetching gitfs remote 'https://gitlab.com/<redacted>.git': too many redirects or authentication replays
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/salt/utils/gitfs.py", line 1865, in _fetch
    fetch_results = origin.fetch(**fetch_kwargs)
  File "/usr/lib/python3/dist-packages/pygit2/remote.py", line 423, in fetch
    check_error(err)
  File "/usr/lib/python3/dist-packages/pygit2/errors.py", line 61, in check_error
    raise GitError(message)
_pygit2.GitError: too many redirects or authentication replays
```
```
git clone https://$(salt-run sdb.get sdb://redis/gitlab/user):$(salt-run sdb.get sdb://redis/gitlab/password)@gitlab.com/<redacted>.git
warning: You appear to have cloned an empty repository.
```

### Commits signed with GPG?
No

### CODE REVIEW

File: salt/utils/sdb.py [hunk 1]
```diff
@@ -57,7 +57,13 @@ def sdb_get(uri, opts, utils=None, strict=False):
     query = uri[indx + 1 :]
 
     loaded_db = salt.loader.sdb(opts, fun, utils=utils)
-    return loaded_db[fun](query, profile=profile)
+    sdb_value = loaded_db[fun](query, profile=profile)
```
Comment (Ch3LL) [inline salt/utils/sdb.py:60]: Why is it not always returning a string? My concerns is there might be use cases where someone would want a bytes value.

**Comment by jynolen** [comment]
IMO is that you probably don't want a byte value within configuration files as it's the case right now.
In another scénario like state your can still use jinja to override string to bytes.
Still my opinion but string scénario is the more probable one.

**Comment by jynolen** [comment]
Hello any updates ?

**Comment by Ch3LL** [comment]
Let me reach out to @saltstack/team-core to find someone who is more familiar to see if we are okay with only strings.

**Comment by Ch3LL** [comment]
I spoke with the team, and it does look like sdb expects strings, but would prefer to fix it earlier in the stack. For the redis SDB module, according to their docs: https://github.com/andymccurdy/redis-py

> If all string responses from a client should be decoded, the user can specify decode_responses=True to Redis.__init__. In this case, any Redis command that returns a string type will be decoded with the encoding specified.

I would prefer to take the approach to ensure `decode_responses=True` is passed with the StrictRedis call in the redis sdb module. to fix the issue. In the redis returner, module, engine and token we are also passing `decode_responses` to the `redis.StrictRedis` call.

**Comment by sagetherage** [comment]
@jynolen have you reviewed the changes proposed? Are you willing to still work on this PR and make those changes?

**Comment by jynolen** [comment]
Hello sorry for late response.
Proposed changes is only available for redis.
I'm not sure it's available for others. I take redis as example because it's easier to understand IMO. So sadly I'm not sure it will fit to all sdb modules

**Comment by Ch3LL** [comment]
Do you know of any other sdb modules that are affected?

**Comment by Ch3LL** [comment]
Just following up here. I still think the right approach is to fix it earlier in the stack in the sdb redis module itself. If there are other sdb modules that are also running into this issue we should also fix those issues in their sdb module files. I'm willing to help with this change and help add test coverage. Are you willing to push up the change to the redis module and add tests? If not I can set aside time to do so.

**Comment by Ch3LL** [comment]
Just following up here. Did you see my comment above?

**Comment by jynolen** [comment]
Hello @Ch3LL , if i rephrase your comment.
You want to implement a safe decode in each sdb module individually.
My two cents that it will need to change probably a lot of them because issue is present for at least 3 sdb.
My POV is to done the oposiste way. Fix it in a global way and implement fallback to bytes case per case when needed.

**Comment by Ch3LL** [comment]
Yes, I am suggesting we implement the fix in multiple modules if needed. The reason why I think the fix earlier in the stack is a better approach is we can ensure we are handling the correct data type, when we are further down the stack. If we don't fix it earlier in the stack we have to continue to handle this data conversion later in the stack if the redis sdb code ever ends up being used in other parts of Salt's code. 

I can get some others opinions here as well. ping @dwoz or @waynew any opinions here?

**Comment by waynew** [comment]
@Ch3LL hit it right on the head here - I saw the exact same issues on https://github.com/saltstack/salt/pull/61229 sdb was one of the next things I had targeted, along with pillars at least.

Basically what's happening is that behind the scenes there's probably a setting for sqlite in the sdb that says, "oh yeah you should return text as strings" -- for a stopgap fix, that would be a good one. If we had an unlimited amount of time, we should have complete sdb tests, it's just not been surfaced back up to the top of my priority list :upside_down_face:

**Comment by Ch3LL** [comment]
Thanks for confirming. @jynolen you willing to update the redis sdb module with these changes or do you want to close the PR? I can open an issue and fix this in the next scheduled release in my own time if you want to close the PR.

**Comment by Ch3LL** [comment]
I'm going to close this PR and make sure https://github.com/saltstack/salt/issues/58533  will remain open. It is slated to be fixed in 3006.

**Pull request closed without merging**